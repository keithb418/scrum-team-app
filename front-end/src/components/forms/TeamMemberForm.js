import React, { Component } from "react";
import { Form, FormGroup, FormControl, ControlLabel, Button, Checkbox, Radio } from "react-bootstrap";
import FontAwesome from "react-fontawesome";
import PropTypes from "prop-types";

import SelectRole from "./SelectRole";
import SelectTeam from "./SelectTeam";
import AddSkills from "./AddSkills";
import CancelButton from "./CancelButton";

import { isEmailValid } from "../../utils/stringHelpers"

class TeamMemberForm extends Component {
  constructor (props) {
    super(props);
    this.state = {
      name: props.teamMember ? props.teamMember.name : "",
      email: props.teamMember ? props.teamMember.email : "",
      team: props.teamMember ? props.teamMember.team : "",
      teamLead: props.teamMember ? props.teamMember.teamLead : false,
      experience: props.teamMember ? props.teamMember.experience : "",
      role: props.teamMember ? props.teamMember.role : "",
      skills: props.teamMember ? props.teamMember.skills : [],
      title: props.title
    };
  }

  onInputChange = (e) => {
    const target = e.target;
    const value = target.type === "checkbox" ? target.checked : target.value;
    const name = target.name;

    this.setState({
      [name]: value
    });
  }

  onSkillsChange = (skills) => {
    this.setState({
      skills
    });
  }

  getValidationState = () => {
    const nameLength = this.state.name.length;
    const emailValid = isEmailValid(this.state.email);
    const teamSelect = this.state.team;

    if (nameLength > 0 && emailValid && teamSelect !== "Select Team") {
      return "success";
    } else {
      return "error";
    }
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const { name, email, team, teamLead, role, skills, experience } = this.state;
    const validationState = this.getValidationState();

    if (validationState === "error") {
      return false;
    } else {
      this.props.onSubmit({ 
        name, 
        email, 
        team, 
        teamLead, 
        role, 
        skills, 
        experience
      });
    }
  }

  render () {
    const isDisabled = this.getValidationState() === "error" ? this.state.isDisabled = true : this.state.isDisabled = false;

    return(
      <div className="row">
        <div className="form-section col-md-6 offset-md-3 col-xs-12">
          <div className="panel-heading">
            <h3 className="panel-title">{this.props.title}</h3>
          </div>
          <hr />
          <Form onSubmit={this.handleSubmit}>
            <ControlLabel>Name<sup><i className="fa fa-asterisk required"></i></sup></ControlLabel>
            <FormGroup controlId="name">
              <FormControl
                type="text"
                name="name"
                placeholder="Enter your name"
                value={this.state.name}
                onChange={this.onInputChange}
                autoFocus
                required />
            </FormGroup>

            <ControlLabel>Email<sup><i className="fa fa-asterisk required"></i></sup></ControlLabel>
            <FormGroup controlId="email">
              <FormControl
                type="email"
                name="email"
                placeholder="Enter your email"
                value={this.state.email}
                onChange={this.onInputChange} />
            </FormGroup>

            <FormGroup controlId="teamLead">
              <span className="right-spacer">Team Lead</span>
              <Checkbox
                name="teamLead"
                value={this.state.teamLead}
                checked={this.state.teamLead ? "checked" : ""}
                onChange={this.onInputChange}
                inline>
              </Checkbox>
            </FormGroup>

            <ControlLabel>Experience</ControlLabel>
            <FormGroup controlId="experience">
              <Radio
                className="right-spacer"
                name="experience"
                value="Senior Level"
                checked={this.state.experience === "Senior Level"}
                onChange={this.onInputChange}
                inline>
                <span className="left-spacer">Senior Level</span>
              </Radio>
              <Radio
                className="right-spacer"
                name="experience"
                value="Mid Level"
                checked={this.state.experience === "Mid Level"}
                onChange={this.onInputChange}
                inline>
                <span className="left-spacer">Mid Level</span>
              </Radio>
              <Radio
                className="right-spacer"
                name="experience"
                value="Junior Level"
                checked={this.state.experience === "Junior Level"}
                onChange={this.onInputChange}
                inline>
                <span className="left-spacer">Junior Level</span>
              </Radio>
              <Radio
                className="right-spacer"
                name="experience"
                value="Level not applicable"
                checked={this.state.experience === "Level not applicable"}
                onChange={this.onInputChange}
                inline>
                <span className="left-spacer">Level not applicable</span>
              </Radio>
            </FormGroup>

            <SelectTeam
              name="team"
              teams={this.props.teams}
              selected={this.state.team}
              onSelect={this.onInputChange} />

            <SelectRole
              roles={this.props.roles}
              selected={this.state.role}
              onSelect={this.onInputChange} />

            <AddSkills
              id="add-skills"
              onChange={this.onSkillsChange}
              skills={this.state.skills} />

            <p>Please enter one skill at a time</p>
            <Button
              type="submit"
              bsStyle="primary"
              disabled={isDisabled}
            >
              {this.props.title}
            </Button>
            <Button type="reset">Reset</Button>
            <CancelButton />
          </Form>
        </div>
      </div>
    );
  }
};

TeamMemberForm.propTypes = {
  createTeamMember: PropTypes.func,
  roles: PropTypes.array,
  teams: PropTypes.array,
  teamMember: PropTypes.object
};

export default TeamMemberForm;