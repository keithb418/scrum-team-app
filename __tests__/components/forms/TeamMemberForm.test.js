import React from 'react';
import { shallow } from 'enzyme';
import TeamMemberForm from '../../../front-end/src/components/forms/TeamMemberForm';
import teamMembersData from '../../fixtures/teamMembers';
import teamsData from '../../fixtures/teams';
import rolesData from '../../fixtures/roles';

test('should render TeamMemberForm correctly', () => {
    const wrapper = shallow(
        <TeamMemberForm
            title="Test Team Form"
            teamMember={teamMembersData[0]}
            teams={teamsData}
            roles={rolesData}/>
    );
    expect(wrapper).toMatchSnapshot()
});

test('should render invalid TeamMemberForm correctly', () => {
    const getValidationState = jest.fn().mockImplementation(() => "error");
    const wrapper = shallow(<TeamMemberForm/>);
    wrapper.find('Form').simulate('submit', {
        preventDefault: () => { }
    });
    expect(wrapper.state('isDisabled')).toBe(true);
    expect(wrapper.find('Button').at(0).props().disabled).toBe(true);
    expect(wrapper).toMatchSnapshot();
});

test('should render invalid TeamMemberForm correctly', () => {
    const wrapper = shallow(<TeamMemberForm/>);
    const name = "name";
    const value = "Team Member 1";
    wrapper.find('FormControl').at(0).simulate('change', {
        target: { name, value }
    });
    expect(wrapper.state('name')).toBe(value);
});

test('should render invalid TeamMemberForm correctly', () => {
    const wrapper = shallow(<TeamMemberForm/>);
    const name = "email";
    const value = "test123@mail.com";
    wrapper.find('FormControl').at(1).simulate('change', {
        target: { name, value }
    });
    expect(wrapper.state('email')).toBe(value);
});

test('should render invalid TeamMemberForm correctly', () => {
    const wrapper = shallow(<TeamMemberForm/>);
    const name = "teamLead";
    const checked = false;
    const value = false
    wrapper.find('Checkbox').simulate('change', {
        target: { name, checked, value }
    });
    expect(wrapper.state('teamLead')).toBe(checked);
});

test('should render invalid TeamMemberForm correctly', () => {
    const wrapper = shallow(<TeamMemberForm/>);
    const name = "experience";
    const value = "Senior Level";
    wrapper.find('Radio').at(0).simulate('change', {
        target: { name, value }
    });
    expect(wrapper.state('experience')).toBe(value);
});

test('should render invalid TeamMemberForm correctly', () => {
    const wrapper = shallow(<TeamMemberForm/>);
    const name = "experience"
    const value = "Mid Level";
    wrapper.find('Radio').at(1).simulate('change', {
        target: { name, value }
    });
    expect(wrapper.state('experience')).toBe(value);
});

test('should render invalid TeamMemberForm correctly', () => {
    const wrapper = shallow(<TeamMemberForm/>);
    const name = "experience"
    const value = "Junior Level";
    wrapper.find('Radio').at(2).simulate('change', {
        target: { name, value }
    });
    expect(wrapper.state('experience')).toBe(value);
});

test('should render invalid TeamMemberForm correctly', () => {
    const wrapper = shallow(<TeamMemberForm/>);
    const name = "experience"
    const value = "Level not applicable";
    wrapper.find('Radio').at(3).simulate('change', {
        target: { name, value }
    });
    expect(wrapper.state('experience')).toBe(value);
});

test('should set select team', () => {
    const value = teamMembersData[0].team;
    const wrapper = shallow(<TeamMemberForm teamMember={teamMembersData[0]}/>);
    expect(wrapper.state('team')).toBe(value);
});

test('should set select roles', () => {
    const value = teamMembersData[2].role;
    const wrapper = shallow(<TeamMemberForm teamMember={teamMembersData[2]}/>);
    expect(wrapper.state('role')).toBe(value);
});

test('should call onSubmit prop for valid form submission', () => {
    const onSubmit = jest.fn();
    const getValidationState = jest.fn().mockImplementation(() => "success");
    const wrapper = shallow(
        <TeamMemberForm
            title="Test Team Form"
            teamMember={teamMembersData[0]}
            onSubmit={onSubmit}/>
    );
    wrapper.find('Form').simulate('submit', {
        preventDefault: () => { }
    });
    expect(wrapper.state('isDisabled')).toBe(false);
    expect(wrapper.find('Button').at(0).props().disabled).toBe(false);
    expect(onSubmit).toHaveBeenLastCalledWith({
        name: teamMembersData[0].name,
        email: teamMembersData[0].email,
        team: teamMembersData[0].team,
        teamLead: teamMembersData[0].teamLead,
        experience: teamMembersData[0].experience,
        role: teamMembersData[0].role,
        skills: teamMembersData[0].skills,
    });
});


