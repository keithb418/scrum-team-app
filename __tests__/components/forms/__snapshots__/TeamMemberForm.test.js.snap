// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TeamMemberForm correctly 1`] = `
<div
  className="row"
>
  <div
    className="form-section col-md-6 offset-md-3 col-xs-12"
  >
    <div
      className="panel-heading"
    >
      <h3
        className="panel-title"
      >
        Test Team Form
      </h3>
    </div>
    <hr />
    <Form
      bsClass="form"
      componentClass="form"
      horizontal={false}
      inline={false}
      onSubmit={[Function]}
    >
      <ControlLabel
        bsClass="control-label"
        srOnly={false}
      >
        Name
        <sup>
          <i
            className="fa fa-asterisk required"
          />
        </sup>
      </ControlLabel>
      <FormGroup
        bsClass="form-group"
        controlId="name"
      >
        <FormControl
          autoFocus={true}
          bsClass="form-control"
          componentClass="input"
          name="name"
          onChange={[Function]}
          placeholder="Enter your name"
          required={true}
          type="text"
          value="Marc"
        />
      </FormGroup>
      <ControlLabel
        bsClass="control-label"
        srOnly={false}
      >
        Email
        <sup>
          <i
            className="fa fa-asterisk required"
          />
        </sup>
      </ControlLabel>
      <FormGroup
        bsClass="form-group"
        controlId="email"
      >
        <FormControl
          bsClass="form-control"
          componentClass="input"
          name="email"
          onChange={[Function]}
          placeholder="Enter your email"
          type="email"
          value="mtest@mail.com"
        />
      </FormGroup>
      <FormGroup
        bsClass="form-group"
        controlId="teamLead"
      >
        <span
          className="right-spacer"
        >
          Team Lead
        </span>
        <Checkbox
          bsClass="checkbox"
          checked=""
          disabled={false}
          inline={true}
          name="teamLead"
          onChange={[Function]}
          title=""
          value={false}
        />
      </FormGroup>
      <ControlLabel
        bsClass="control-label"
        srOnly={false}
      >
        Experience
      </ControlLabel>
      <FormGroup
        bsClass="form-group"
        controlId="experience"
      >
        <Radio
          bsClass="radio"
          checked={false}
          className="right-spacer"
          disabled={false}
          inline={true}
          name="experience"
          onChange={[Function]}
          title=""
          value="Senior Level"
        >
          <span
            className="left-spacer"
          >
            Senior Level
          </span>
        </Radio>
        <Radio
          bsClass="radio"
          checked={false}
          className="right-spacer"
          disabled={false}
          inline={true}
          name="experience"
          onChange={[Function]}
          title=""
          value="Mid Level"
        >
          <span
            className="left-spacer"
          >
            Mid Level
          </span>
        </Radio>
        <Radio
          bsClass="radio"
          checked={false}
          className="right-spacer"
          disabled={false}
          inline={true}
          name="experience"
          onChange={[Function]}
          title=""
          value="Junior Level"
        >
          <span
            className="left-spacer"
          >
            Junior Level
          </span>
        </Radio>
        <Radio
          bsClass="radio"
          checked={false}
          className="right-spacer"
          disabled={false}
          inline={true}
          name="experience"
          onChange={[Function]}
          title=""
          value="Level not applicable"
        >
          <span
            className="left-spacer"
          >
            Level not applicable
          </span>
        </Radio>
      </FormGroup>
      <SelectTeam
        name="team"
        onSelect={[Function]}
        selected="1abc"
        teams={
          Array [
            Object {
              "_id": "1",
              "name": "Team 1",
            },
            Object {
              "_id": "2",
              "name": "Team 2",
            },
          ]
        }
      />
      <SelectRole
        onSelect={[Function]}
        roles={
          Array [
            Object {
              "_id": "1",
              "name": "Front-End Developer",
            },
            Object {
              "_id": "2",
              "name": "Back-End Developer",
            },
            Object {
              "_id": "3",
              "name": "Full-Stack Developer",
            },
            Object {
              "_id": "4",
              "name": "Manual Tester",
            },
            Object {
              "_id": "5",
              "name": "Automation Tester",
            },
          ]
        }
        selected="Fullstack Developer"
      />
      <AddSkills
        id="add-skills"
        onChange={[Function]}
        skills={
          Array [
            "react",
          ]
        }
      />
      <p>
        Please enter one skill at a time
      </p>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="primary"
        disabled={false}
        type="submit"
      >
        Test Team Form
      </Button>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        disabled={false}
        type="reset"
      >
        Reset
      </Button>
      <CancelButton />
    </Form>
  </div>
</div>
`;

exports[`should render invalid TeamMemberForm correctly 1`] = `
<div
  className="row"
>
  <div
    className="form-section col-md-6 offset-md-3 col-xs-12"
  >
    <div
      className="panel-heading"
    >
      <h3
        className="panel-title"
      />
    </div>
    <hr />
    <Form
      bsClass="form"
      componentClass="form"
      horizontal={false}
      inline={false}
      onSubmit={[Function]}
    >
      <ControlLabel
        bsClass="control-label"
        srOnly={false}
      >
        Name
        <sup>
          <i
            className="fa fa-asterisk required"
          />
        </sup>
      </ControlLabel>
      <FormGroup
        bsClass="form-group"
        controlId="name"
      >
        <FormControl
          autoFocus={true}
          bsClass="form-control"
          componentClass="input"
          name="name"
          onChange={[Function]}
          placeholder="Enter your name"
          required={true}
          type="text"
          value=""
        />
      </FormGroup>
      <ControlLabel
        bsClass="control-label"
        srOnly={false}
      >
        Email
        <sup>
          <i
            className="fa fa-asterisk required"
          />
        </sup>
      </ControlLabel>
      <FormGroup
        bsClass="form-group"
        controlId="email"
      >
        <FormControl
          bsClass="form-control"
          componentClass="input"
          name="email"
          onChange={[Function]}
          placeholder="Enter your email"
          type="email"
          value=""
        />
      </FormGroup>
      <FormGroup
        bsClass="form-group"
        controlId="teamLead"
      >
        <span
          className="right-spacer"
        >
          Team Lead
        </span>
        <Checkbox
          bsClass="checkbox"
          checked=""
          disabled={false}
          inline={true}
          name="teamLead"
          onChange={[Function]}
          title=""
          value={false}
        />
      </FormGroup>
      <ControlLabel
        bsClass="control-label"
        srOnly={false}
      >
        Experience
      </ControlLabel>
      <FormGroup
        bsClass="form-group"
        controlId="experience"
      >
        <Radio
          bsClass="radio"
          checked={false}
          className="right-spacer"
          disabled={false}
          inline={true}
          name="experience"
          onChange={[Function]}
          title=""
          value="Senior Level"
        >
          <span
            className="left-spacer"
          >
            Senior Level
          </span>
        </Radio>
        <Radio
          bsClass="radio"
          checked={false}
          className="right-spacer"
          disabled={false}
          inline={true}
          name="experience"
          onChange={[Function]}
          title=""
          value="Mid Level"
        >
          <span
            className="left-spacer"
          >
            Mid Level
          </span>
        </Radio>
        <Radio
          bsClass="radio"
          checked={false}
          className="right-spacer"
          disabled={false}
          inline={true}
          name="experience"
          onChange={[Function]}
          title=""
          value="Junior Level"
        >
          <span
            className="left-spacer"
          >
            Junior Level
          </span>
        </Radio>
        <Radio
          bsClass="radio"
          checked={false}
          className="right-spacer"
          disabled={false}
          inline={true}
          name="experience"
          onChange={[Function]}
          title=""
          value="Level not applicable"
        >
          <span
            className="left-spacer"
          >
            Level not applicable
          </span>
        </Radio>
      </FormGroup>
      <SelectTeam
        name="team"
        onSelect={[Function]}
        selected=""
      />
      <SelectRole
        onSelect={[Function]}
        selected=""
      />
      <AddSkills
        id="add-skills"
        onChange={[Function]}
        skills={Array []}
      />
      <p>
        Please enter one skill at a time
      </p>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="primary"
        disabled={true}
        type="submit"
      />
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        disabled={false}
        type="reset"
      >
        Reset
      </Button>
      <CancelButton />
    </Form>
  </div>
</div>
`;
